python/src/iceberg/io/s3.py

```diff
@@ -0,0 +1,185 @@
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from typing import Union
+from urllib.parse import urlparse
+
+import boto3
+from botocore.errorfactory import ClientError
+
+from iceberg.io.base import FileIO, InputFile, OutputFile
+
+
+class S3InputFile(InputFile):
+    def __init__(self, location: str):
+        super().__init__(location=location)
+        self._s3_client = None
+
+    def __len__(self, s3_client=None):
+        s3 = self._create_s3_session() if not s3_client else s3_client
+        s3_url = urlparse(self.location)
+        resource = s3.head_object(Bucket=s3_url.netloc, Key=s3_url.path.lstrip("/"))
+        return resource["ContentLength"]
+
+    def _create_s3_session(self):
+        return boto3.Session().client("s3")
+
+    def exists(self, s3_client=None):
+        s3 = self._create_s3_session() if not s3_client else s3_client
+        s3_url = urlparse(self.location)
+        try:
+            s3.head_object(Bucket=s3_url.netloc, Key=s3_url.path.lstrip("/"))
+        except ClientError as e:
+            if "Not Found" in str(e):
+                return False
+            else:
+                raise
+
+        return True
+
+    def __call__(self, s3_client=None):
+        """Allows injecting an s3 client before the context is opened"""
+        self._s3_client = s3_client
+        return self
+
+    def __enter__(self):
+        """Enter context for S3InputFile
+
+        This sets a botocore.response.StreamingBody instance for the
+        location to self.input_stream. If an `s3_client` is not provided through
+        a call during context open, a boto3 session using the environment AWS
+        credentials will be created.
+
+        Example:
+            >>> input_file = S3InputFile(location="s3://foo/bar.json")
+                with input_file as f:
+                    data = input_file.input_stream.read()
+
+            >>> import boto3
+                AWS_ACCESS_KEY_ID = "..."
+                AWS_SECRET_ACCESS_KEY = "..."
+                s3_client = boto3.Session(
+                    aws_access_key_id=AWS_ACCESS_KEY_ID,
+                    aws_secret_access_key=AWS_SECRET_ACCESS_KEY,
+                ).client("s3")
+                input_file = S3InputFile(location="s3://baz/qux.json")
+                with input_file(s3_client=s3_client) as f:
+                    data = input_file.input_stream.read()
+
+
+        """
+        super().__enter__()
```

Comment:
I'd avoid calling super. We know it doesn't do anything (assuming we remove logging).

Reply:
Removed all logging and super calls except in the init.